libs/p2p-crypto/src/main/kotlin/net/corda/p2p/crypto/protocol/api/CertificateValidator.kt

```diff
@@ -0,0 +1,105 @@
+import java.security.cert.CertPathValidatorException
+import java.security.cert.CertificateFactory
+import java.security.cert.PKIXBuilderParameters
+import java.security.cert.PKIXRevocationChecker
+import java.security.cert.X509CertSelector
+import java.security.cert.X509Certificate
+
+class CertificateValidator(
+    private val revocationCheckMode: RevocationCheckMode,
+    private val trustStore: KeyStore,
+    private val certPathValidator: CertPathValidator = CertPathValidator.getInstance(certificateAlgorithm),
+    private val certificateFactory: CertificateFactory = CertificateFactory.getInstance(certificateFactoryType),
+) {
+
+    private companion object {
+        const val certificateAlgorithm = "PKIX"
+        const val certificateFactoryType = "X.509"
+        const val digitalSignatureBit = 0
+        val logger = contextLogger()
+    }
+
+    @Suppress("ThrowsCount")
+    fun validate(cert: List<String>, expectedX500Name: MemberX500Name) {
+        val certificateChain = certificateFactory.generateCertPath(cert.map { pemCertificate ->
+            ByteArrayInputStream(pemCertificate.toByteArray()).use {
+                certificateFactory.generateCertificate(it)
+            }
+        })
+        //By convention, the certificates in a CertPath object of type X.509 are ordered starting with the target certificate
+        //and ending with a certificate issued by the trust anchor. So we check the subjectX500Principal of the first certificate
+        //matches the x500Name of the peer's identity.
+        val x509RootCert = (certificateChain.certificates.firstOrNull() as? X509Certificate) ?:
+            throw InvalidPeerCertificate("Root session certificate is not an X509 certificate.")
+
+        validateX500NameMatches(x509RootCert, expectedX500Name)
+        validateKeyUsage(x509RootCert)
+        validateCertPath(certificateChain)
+    }
+
+    private fun validateX500NameMatches(certificate: X509Certificate, expectedX500Name: MemberX500Name) {
+        val x500PrincipalFromCert = certificate.subjectX500Principal
+        val x500NameFromCert = try {
+            MemberX500Name.build(x500PrincipalFromCert)
+        } catch (exception: IllegalArgumentException) {
+            throw InvalidPeerCertificate(
+                "X500 principle in root session certificate ($x500PrincipalFromCert) is not a valid corda X500 Name.",
+                certificate
+            )
+        }
+        if (x500NameFromCert != expectedX500Name) {
+            throw InvalidPeerCertificate(
+                "X500 principle in root session certificate ($x500NameFromCert) is different from expected ($expectedX500Name).",
+                certificate
+            )
+        }
+    }
+
+    private fun validateKeyUsage(certificate: X509Certificate) {
+        if (!certificate.keyUsage[digitalSignatureBit]) {
+            throw InvalidPeerCertificate("The key usages extension of the root session certificate does not contain " +
```

Comment:
```suggestion
            throw InvalidPeerCertificate("The key usages extension of the session certificate does not contain " +
```
(related to my other comment on this being the leaf, not the root cert)